File: plugins/secure_storage/android/src/main/kotlin/kg/ae/cv_scan/secure_storage/SecureStoragePlugin.kt
package kg.ae.cv_scan.secure_storage

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import java.util.concurrent.Executors

/**
 * Keystore-backed implementation of the Pigeon [SecureStorageApi].
 *
 * Crypto and disk I/O run on a single background thread (serializing Keystore
 * access, which is not thread-safe), and every reply is posted back to the main
 * thread as the platform-channel contract requires.
 */
class SecureStoragePlugin :
    FlutterPlugin,
    SecureStorageApi {

    private lateinit var prefs: SharedPreferences
    private lateinit var cipher: KeystoreCipher

    private val worker = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        prefs = binding.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        cipher = KeystoreCipher(KEY_ALIAS)
        SecureStorageApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        SecureStorageApi.setUp(binding.binaryMessenger, null)
    }

    override fun read(key: String, callback: (Result<String?>) -> Unit) = dispatch(callback) {
        prefs.getString(key, null)?.let(cipher::decrypt)
    }

    override fun write(key: String, value: String, callback: (Result<Unit>) -> Unit) = dispatch(callback) {
        prefs.edit().putString(key, cipher.encrypt(value)).commit()
    }

    override fun delete(key: String, callback: (Result<Unit>) -> Unit) = dispatch(callback) {
        prefs.edit().remove(key).commit()
    }

    override fun deleteAll(callback: (Result<Unit>) -> Unit) = dispatch(callback) {
        prefs.edit().clear().commit()
    }

    override fun containsKey(key: String, callback: (Result<Boolean>) -> Unit) = dispatch(callback) {
        prefs.contains(key)
    }

    private fun <T> dispatch(callback: (Result<T>) -> Unit, action: () -> T) {
        worker.execute {
            val result = runCatching(action)
            main.post { callback(result) }
        }
    }

    private companion object {
        const val PREFS_NAME = "kg.ae.cv_scan.secure_storage"
        const val KEY_ALIAS = "cv_scan_secure_storage_master_key"
    }
}

File: plugins/secure_storage/ios/cv_scan_secure_storage/Sources/cv_scan_secure_storage/SecureStoragePlugin.swift
import Flutter
import UIKit

/// Keychain-backed implementation of the Pigeon `SecureStorageApi`.
///
/// Keychain operations are synchronous and run on the platform thread the
/// channel handler is invoked on; failures surface to Dart as a `PigeonError`.
public class SecureStoragePlugin: NSObject, FlutterPlugin, SecureStorageApi {
  private let store = KeychainStore(
    service: "kg.ae.cv_scan.secure_storage",
    accessibility: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
  )

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = SecureStoragePlugin()
    SecureStorageApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
  }

  func read(key: String, completion: @escaping (Result<String?, Error>) -> Void) {
    completion(runCatching { try store.read(key: key) })
  }

  func write(key: String, value: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(runCatching { try store.write(key: key, value: value) })
  }

  func delete(key: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(runCatching { try store.delete(key: key) })
  }

  func deleteAll(completion: @escaping (Result<Void, Error>) -> Void) {
    completion(runCatching { try store.deleteAll() })
  }

  func containsKey(key: String, completion: @escaping (Result<Bool, Error>) -> Void) {
    completion(runCatching { try store.contains(key: key) })
  }

  private func runCatching<T>(_ action: () throws -> T) -> Result<T, Error> {
    do {
      return .success(try action())
    } catch {
      return .failure(error)
    }
  }
}
